src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/AddUsingsCodeActionResolver.cs

```diff
@@ -0,0 +1,243 @@
+                var edits = GenerateSingleUsingEditsInterpolated(codeDocument, codeDocumentIdentifier, actionParams.Namespace, usingDirectives);
+                documentChanges.Add(edits);
+            }
+            else
+            {
+                // Just throw them at the top
+                var edits = GenerateSingleUsingEditsAtTop(codeDocument, codeDocumentIdentifier, actionParams.Namespace);
+                documentChanges.Add(edits);
+            }
+
+            return new WorkspaceEdit()
+            {
+                DocumentChanges = documentChanges
+            };
+        }
+
+        private static WorkspaceEditDocumentChange GenerateSingleUsingEditsInterpolated(
+            RazorCodeDocument codeDocument,
+            VersionedTextDocumentIdentifier codeDocumentIdentifier,
+            string @namespace,
+            List<RazorUsingDirective> existingUsingDirectives)
+        {
+            var edits = new List<TextEdit>();
+            var newText = $"@using {@namespace}{Environment.NewLine}";
+
+            foreach (var usingDirective in existingUsingDirectives)
+            {
+                // Skip System directives; if they're at the top we don't want to insert before them
+                var usingDirectiveNamespace = usingDirective.Statement.ParsedNamespace;
+                if (usingDirectiveNamespace.StartsWith("System"))
+                {
+                    continue;
+                }
+
+                if (@namespace.CompareTo(usingDirectiveNamespace) < 0)
+                {
+                    var usingDirectiveLineIndex = codeDocument.Source.Lines.GetLocation(usingDirective.Node.Span.Start).LineIndex;
+                    var head = new Position(usingDirectiveLineIndex, 0);
+                    var edit = new TextEdit() { Range = new Range(head, head), NewText = newText };
+                    edits.Add(edit);
+                    break;
+                }
+            }
+
+            if (edits.Count == 0)
+            {
+                var endIndex = existingUsingDirectives.Last().Node.Span.End;
+                var lineIndex = GetLineIndexOrEnd(codeDocument, endIndex - 1) + 1;
+                var head = new Position(lineIndex, 0);
+                var edit = new TextEdit() { Range = new Range(head, head), NewText = newText };
+                edits.Add(edit);
+            }
+
+            return new WorkspaceEditDocumentChange(new TextDocumentEdit()
+            {
+                TextDocument = codeDocumentIdentifier,
+                Edits = edits,
+            });
+        }
+
+        private static WorkspaceEditDocumentChange GenerateSingleUsingEditsAtTop(
+            RazorCodeDocument codeDocument,
+            VersionedTextDocumentIdentifier codeDocumentIdentifier,
+            string @namespace)
+        {
+            // If we don't have usings, insert after the last namespace or page directive, which ever comes later
+            var head = new Position(1, 0);
```

Comment:
Would this work well for documents that look like this? Notice the empty line at the top.
![image](https://user-images.githubusercontent.com/1579269/87706609-a6edbd00-c754-11ea-9401-36a8b8c32a86.png)

Reply:
Don't worry, that's just how the control flow works. It defaults to the top but if it finds an @page or @namespace it will replace the cursor with the following line.

Reply:
Ah yeah, missed that it includes Page as well. 👍